### PR TITLE
PUBDEV-6367 - Overridable non-data line markers

### PR DESCRIPTION
**Pull request by Pscheidl**
https://0xdata.atlassian.net/browse/PUBDEV-6367

By default, some parsers are skipping lines beginning with some characters. Those defaults are implemented per each parser. The main point of this PR is to fulfill all the following conditions:

1. Make the defaults overridable,
1. Preserve the original behavior(defaults) when non-data line markers are not specified.

As a positive side-effect of this API, I made `CsvParser` and `ArffParser` use the new generic methods passing the whole `ParseSetup` object (reference to it of course), instead of pin-pointing only the necessary stuff. This old method has even been deprecated by @michalkurka some time ago. :)

### CODE REVIEW

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 1]
```diff
@@ -166,15 +172,36 @@ public ParseSetup(ParseSetupV3 ps) {
...
+        domains, naStrings, data, new ParseWriter.ParseErr[0], FileVec.DFLT_CHUNK_SIZE, nonDataLineMarkers);
   }
+
+  /**
+   * Create a ParseSetup with all parameters except chunk size.
+   *
+   * Typically used by file type parsers for returning final valid results
+   * _chunk_size will be set later using results from all files.
+   */
+  public ParseSetup(ParserInfo parseType, byte sep, boolean singleQuotes, int checkHeader,
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: this is becoming a constructor hell :)

File: h2o-py/h2o/h2o.py [hunk 2]
```diff
@@ -591,6 +593,7 @@ def parse_setup(raw_frames, destination_frame=None, header=0, separator=None, co
     :param na_strings: A list of strings, or a list of lists of strings (one list per column), or a dictionary
         of column names to strings which are to be interpreted as missing values.
     :param skipped_columns: an integer lists of column indices to skip and not parsed into the final frame from the import file.
+    :param non_data_line_markers: A string containing custom characters to be treated as non-data line markers
```
Comment (michalkurka) [inline h2o-py/h2o/h2o.py]: this is confusing, isn't it? 

I would expect that `non_data_line_markers = None` means there are no such markers, not that the default will be applied

File: h2o-core/src/main/java/water/parser/CsvParser.java [hunk 3]
```diff
@@ -18,14 +18,19 @@
...
 
-  CsvParser( ParseSetup ps, Key jobKey ) { super(ps, jobKey); }
+  CsvParser( ParseSetup ps, Key jobKey ) {
+    super(ps, jobKey);
+    if(_setup._nonDataLineMarkers == null) _setup._nonDataLineMarkers = nonDataLineMarkersDefault();
+  }
 
-  protected byte[] nonDataLineMarkers() {
-    return NON_DATA_LINE_MARKERS;
+  protected byte[] nonDataLineMarkersDefault() {
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/parser/CsvParser.java]: this method is useless I think

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 1 above)
Comment (honzasterba) [reply h2o-core/src/main/java/water/parser/ParseSetup.java]: builder?

File: h2o-py/h2o/h2o.py (hunk 2 above)
Comment (honzasterba) [reply h2o-py/h2o/h2o.py]: it also may be confusing what a non-data-line-marker is, I would suggest rewording, to something like:

If a line in imported file starts with any character in given string it will be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used.

Comment (Pscheidl) [reply h2o-py/h2o/h2o.py]: Yes, it is confusing, but as someone hard-coded these values earlier, it is harder to change it right now. So, should we introduce a new default being handed over ? Some special combination of chars, like `DEFAULT` or something like that ? Seems also confusing to the users.

Best solution would be to drop the default behavior.

File: h2o-core/src/main/java/water/parser/CsvParser.java (hunk 3 above)
Comment (Pscheidl) [reply h2o-core/src/main/java/water/parser/CsvParser.java]: It is not. This behavior is overridden in `ARFFParser`.

Comment (Pscheidl) [reply h2o-core/src/main/java/water/parser/CsvParser.java]: Unfortunately, it is not. The behavior is overridden in `ARFFParser`, which calls CSVParser's constructor in it's own constructor. Each class has different defaults to provide.

File: h2o-py/h2o/h2o.py (hunk 2 above)
Comment (Pscheidl) [reply h2o-py/h2o/h2o.py]: One solution to keep the old behavior and to make it less cofusing is to named the parameter `custom_non_data_line_markers`.

Comment (Pscheidl) [reply h2o-py/h2o/h2o.py]: See commit: https://github.com/h2oai/h2o-3/pull/3380/commits/827331e88c90f0317ec303b23a80aaf5fa9a6470

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 1 above)
Comment (Pscheidl) [reply h2o-core/src/main/java/water/parser/ParseSetup.java]: Thanks for pointing it out. Turns out the constructor injection can be "eased" by deleting some of the older ones. Those were not marked as unused code, but the ones that remained after commit https://github.com/h2oai/h2o-3/pull/3380/commits/0e610d07febddbc27082c31202d9bbbc4ddcbe7b are completely OK.

**Comment by Pscheidl** [comment]
@michalkurka @honzasterba Please re-review

File: h2o-py/h2o/h2o.py [hunk 4]
```diff
@@ -400,6 +400,7 @@ def import_file(path=None, destination_frame=None, parse=True, header=0, sep=Non
     :param pattern: Character string containing a regular expression to match file(s) in the folder if `path` is a
         directory.
     :param skipped_columns: an integer list of column indices to skip and not parsed into the final frame from the import file.
+    :param non_data_line_markers: If a line in imported file starts with any character in given string it will be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used
```
Comment (honzasterba) [inline h2o-py/h2o/h2o.py]: I think this is not correct, correct is:

If a line in imported file starts with any character in given string it will NOT be imported.

File: h2o-py/h2o/h2o.py [hunk 5]
```diff
@@ -555,7 +557,7 @@ def import_sql_select(connection_url, select_query, username, password, optimize
 
 
 def parse_setup(raw_frames, destination_frame=None, header=0, separator=None, column_names=None,
-                column_types=None, na_strings=None, skipped_columns=None):
+                column_types=None, na_strings=None, skipped_columns=None, custom_non_data_line_markers = None):
```
Comment (honzasterba) [inline h2o-py/h2o/h2o.py]: does not match formatting of previous args

File: h2o-py/h2o/h2o.py [hunk 6]
```diff
@@ -591,6 +593,7 @@ def parse_setup(raw_frames, destination_frame=None, header=0, separator=None, co
     :param na_strings: A list of strings, or a list of lists of strings (one list per column), or a dictionary
         of column names to strings which are to be interpreted as missing values.
     :param skipped_columns: an integer lists of column indices to skip and not parsed into the final frame from the import file.
+    :param custom_non_data_line_markers: If a line in imported file starts with any character in given string it will be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used
```
Comment (honzasterba) [inline h2o-py/h2o/h2o.py]: same as above

File: h2o-r/h2o-package/R/import.R [hunk 7]
```diff
@@ -55,6 +55,7 @@
 #' @param decrypt_tool (Optional) Specify a Decryption Tool (key-reference
 #'        acquired by calling \link{h2o.decryptionSetup}.
 #' @param skipped_columns a list of column indices to be skipped during parsing.
+#' @param custom_non_data_line_markers (Optional) If a line in imported file starts with any character in given string it will be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used
```
Comment (honzasterba) [inline h2o-r/h2o-package/R/import.R]: same as above + R does not have None but NULL

File: h2o-r/h2o-package/R/import.R [hunk 8]
```diff
@@ -121,7 +125,8 @@ h2o.importFolder <- function(path, pattern = "", destination_frame = "", parse =
 if(parse) {
     srcKey <- res$destination_frames
     return( h2o.parseRaw(data=.newH2OFrame(op="ImportFolder",id=srcKey,-1,-1),pattern=pattern, destination_frame=destination_frame,
-            header=header, sep=sep, col.names=col.names, col.types=col.types, na.strings=na.strings, decrypt_tool=decrypt_tool, skipped_columns=skipped_columns) )
+            header=header, sep=sep, col.names=col.names, col.types=col.types, na.strings=na.strings, decrypt_tool=decrypt_tool,
+             skipped_columns=skipped_columns, custom_non_data_line_markers = custom_non_data_line_markers) )
```
Comment (honzasterba) [inline h2o-r/h2o-package/R/import.R]: whitespace

File: h2o-r/h2o-package/R/parse.R [hunk 9]
```diff
@@ -30,22 +30,25 @@
 #'        acquired by calling \link{h2o.decryptionSetup}.
 #' @param chunk_size size of chunk of (input) data in bytes
 #' @param skipped_columns a list of column indices to be excluded from parsing
+#' @param custom_non_data_line_markers (Optional) If a line in imported file starts with any character in given string it will be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used
```
Comment (honzasterba) [inline h2o-r/h2o-package/R/parse.R]: same as above

File: h2o-r/h2o-package/R/parse.R [hunk 10]
```diff
@@ -140,6 +147,8 @@ h2o.parseSetup <- function(data, pattern="", destination_frame = "", header = NA
   # begin the setup
   # setup the parse parameters here
   parseSetup.params <- list()
+  
+  if(!is.null(custom_non_data_line_markers)) {parseSetup.params$custom_non_data_line_markers = custom_non_data_line_markers}
```
Comment (honzasterba) [inline h2o-r/h2o-package/R/parse.R]: unnecessary {}, I would either remove the {} or put the statement on a new line

File: h2o-py/h2o/h2o.py (hunk 4 above)
Comment (Pscheidl) [reply h2o-py/h2o/h2o.py]: Sorry, I copy&pasted your proposal without reading it.

File: h2o-py/h2o/h2o.py (hunk 5 above)
Comment (Pscheidl) [reply h2o-py/h2o/h2o.py]: Resolved.

File: h2o-py/h2o/h2o.py (hunk 6 above)
Comment (Pscheidl) [reply h2o-py/h2o/h2o.py]: Resolved.

File: h2o-r/h2o-package/R/import.R (hunk 7 above)
Comment (Pscheidl) [reply h2o-r/h2o-package/R/import.R]: Resolved.

File: h2o-r/h2o-package/R/import.R (hunk 8 above)
Comment (Pscheidl) [reply h2o-r/h2o-package/R/import.R]: Removed

File: h2o-r/h2o-package/R/parse.R (hunk 9 above)
Comment (Pscheidl) [reply h2o-r/h2o-package/R/parse.R]: Resolved, added 'not' plus 'null'

File: h2o-r/h2o-package/R/parse.R (hunk 10 above)
Comment (Pscheidl) [reply h2o-r/h2o-package/R/parse.R]: Style guidelines do not forbid it (https://style.tidyverse.org/syntax.html#indenting - section 2.4.1) . However, the lines was **too long**, so I chose the s econd option.

**Comment by Pscheidl** [comment]
@honzasterba Thanks for the comments again. Is it worthy merging now ?

Gandalf

File: h2o-r/h2o-package/R/parse.R [hunk 11]
```diff
@@ -67,6 +70,9 @@ h2o.parseRaw <- function(data, pattern="", destination_frame = "", header=NA, se
             decrypt_tool = .decrypt_tool_id(parse.params$decrypt_tool),
             skipped_columns = paste0("[", paste(parse.params$skipped_columns, collapse=','), "]")
             )
+   if(!is.null(custom_non_data_line_markers)){
+    parse.params <- append(parse.params,list(custom_non_data_line_markers = custom_non_data_line_markers))
```
Comment (honzasterba) [inline h2o-r/h2o-package/R/parse.R]: whitespace, we use 2 space indent

File: h2o-r/h2o-package/R/parse.R [hunk 12]
```diff
@@ -140,6 +147,10 @@ h2o.parseSetup <- function(data, pattern="", destination_frame = "", header = NA
   # begin the setup
   # setup the parse parameters here
   parseSetup.params <- list()
+  
+  if(!is.null(custom_non_data_line_markers)) {
+   parseSetup.params$custom_non_data_line_markers = custom_non_data_line_markers
```
Comment (honzasterba) [inline h2o-r/h2o-package/R/parse.R]: whitespace, we use 2 space indent

File: h2o-r/h2o-package/R/parse.R [hunk 13]
```diff
@@ -140,6 +147,10 @@ h2o.parseSetup <- function(data, pattern="", destination_frame = "", header = NA
...
+   }
```
Comment (honzasterba) [inline h2o-r/h2o-package/R/parse.R]: whitespace

File: h2o-r/h2o-package/R/parse.R (hunk 11 above)
Comment (Pscheidl) [reply h2o-r/h2o-package/R/parse.R]: This place has +2 spaces indentation. I see no problem here.

File: h2o-r/h2o-package/R/parse.R [hunk 14]
```diff
@@ -127,7 +133,8 @@ h2o.parseRaw <- function(data, pattern="", destination_frame = "", header=NA, se
 #' @seealso \link{h2o.parseRaw}
 #' @export
 h2o.parseSetup <- function(data, pattern="", destination_frame = "", header = NA, sep = "", col.names = NULL, col.types = NULL,
-                           na.strings = NULL, parse_type = NULL, chunk_size = NULL, decrypt_tool = NULL, skipped_columns=NULL) {
+                           na.strings = NULL, parse_type = NULL, chunk_size = NULL, decrypt_tool = NULL, skipped_columns=NULL,
+                           custom_non_data_line_markers=NULL) {
```
Comment (Pscheidl) [inline h2o-r/h2o-package/R/parse.R]: This function has spaces around `=` operator. Some functions have it, some don't. The style is not really unified. To be consistent with function's header, let's change it.

File: h2o-r/h2o-package/R/parse.R (hunk 11 above)
Comment (Pscheidl) [reply h2o-r/h2o-package/R/parse.R]: Re-formatted in R studio. For some reason. GitHub won't show the change 

Snímek z 2019-03-27 14-35-41

Comment (Pscheidl) [reply h2o-r/h2o-package/R/parse.R]: I have no idea why the change is not shown. It has proper indentation after the re-format. Everything else is ok. Everything is pushed.



I even broken it intentionally in https://github.com/h2oai/h2o-3/pull/3380/commits/82c557a520572cbc98a0e238aeb7fc06cb627f25 and then fixed it back in https://github.com/h2oai/h2o-3/pull/3380/commits/cd6352c108608c7c5a9ea48f32106709612fba54 - no luck.

Comment (Pscheidl) [reply h2o-r/h2o-package/R/parse.R]: Great, now it is shown correctly even here, not sure what was wrong with github.

Comment (Pscheidl) [reply h2o-r/h2o-package/R/parse.R]: Here, it looks like `  if(!is.null(custom_non_data_line_markers)){` has three spaces before it, in reality it only has two. See the commit links ^^^

File: h2o-py/h2o/h2o.py [hunk 15]
```diff
@@ -591,6 +593,7 @@ def parse_setup(raw_frames, destination_frame=None, header=0, separator=None, co
     :param na_strings: A list of strings, or a list of lists of strings (one list per column), or a dictionary
         of column names to strings which are to be interpreted as missing values.
     :param skipped_columns: an integer lists of column indices to skip and not parsed into the final frame from the import file.
+    :param custom_non_data_line_markers: If a line in imported file starts with any character in given string it will NOT be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used
```
Comment (ledell) [inline h2o-py/h2o/h2o.py]:
```suggestion
    :param custom_non_data_line_markers: If a line in an imported file starts with any character in the given string, it will be skipped and NOT be imported. An empty string means all lines are imported, None means that default behavior for given format will be used.
```

File: h2o-r/h2o-package/R/parse.R [hunk 16]
```diff
@@ -30,22 +30,25 @@
 #'        acquired by calling \link{h2o.decryptionSetup}.
 #' @param chunk_size size of chunk of (input) data in bytes
 #' @param skipped_columns a list of column indices to be excluded from parsing
+#' @param custom_non_data_line_markers (Optional) If a line in imported file starts with any character in given string it will NOT be imported. Empty string means all lines are imported, NULL means that default behaviour for given format will be used
```
Comment (ledell) [inline h2o-r/h2o-package/R/parse.R]:
```suggestion
#' @param custom_non_data_line_markers (Optional) If a line in an imported file starts with any character in the given string, it will be skipped and NOT be imported. An empty string means all lines are imported, NULL means that default behavior for given format will be used.
```

File: h2o-py/h2o/h2o.py [hunk 17]
```diff
@@ -400,6 +400,7 @@ def import_file(path=None, destination_frame=None, parse=True, header=0, sep=Non
     :param pattern: Character string containing a regular expression to match file(s) in the folder if `path` is a
         directory.
     :param skipped_columns: an integer list of column indices to skip and not parsed into the final frame from the import file.
+    :param custom_non_data_line_markers: If a line in imported file starts with any character in given string it will NOT be imported. Empty string means all lines are imported, None means that default behaviour for given format will be used
```
Comment (ledell) [inline h2o-py/h2o/h2o.py]:
```suggestion
    :param custom_non_data_line_markers: If a line in an imported file starts with any character in the given string, it will be skipped and NOT be imported. An empty string means all lines are imported, None means that default behavior for given format will be used.
```
